app/models/challenge/bylder.rb

```diff
@@ -3,16 +3,36 @@ class Challenge
   class Bylder
     class << self
       def perform(challenger, performance, spot)
-        type = challenge_type_from_spot(spot, performance)
-        users = get_users(challenger, type, spot)
-        challenge = Challenge.new(performance: performance, spot: spot, challenge_type: type, users: users)
-        challenge.user_challenges.each { |uc| uc.spot = uc.user.spot }
+        type = challenge_type_from_spot(spot: spot, performance: performance)
+        users = get_users(challenger: challenger, type: type, spot: spot, performance: performance)
+        challenge = create_or_update(performance: performance, spot: spot, challenge_type: type, users: users)
         challenge
       end
 
       private
 
-      def challenge_type_from_spot(spot, performance)
+      def create_or_update(performance:, spot:, challenge_type:, users:)
+        challenge = find_challenge(performance: performance, spot: spot)
+        if challenge.nil?
+          create(performance: performance, spot: spot, challenge_type: challenge_type, users: users)
+        else
+          update(challenge: challenge, users: users)
+        end
+      end
+
+      def create(performance:, spot:, challenge_type:, users:)
+        challenge = Challenge.new(performance: performance, spot: spot, challenge_type: challenge_type, users: users)
+        challenge.user_challenges.each { |uc| uc.spot = uc.user.spot }
```

Comment:
I think I understand better what you were trying to do here now.

If you want, you can add a `before_create` callback that assigns the spot for you, though some people prefer to be more explicit and avoid those hidden side effects (e.g. in the case where you might need to set a spot explicitly, it would get overwritten unless you either check if there is already a spot or update the user challenge after creation)

Reply:
I don't think there's a case when the spot logic would need to behave differently, but I do prefer to keep it out of a `before_create`